### ISSUE TITLE
ord() and chr() not implemented correctly

### ISSUE THREAD

**Issue by JohnDenker** [BruceSherwood/glowscript#18]
On line  https://github.com/BruceSherwood/glowscript/blob/10103f3e089c3da9b3c7ff8c9406e9462a730882/lib/ace/mode-python.js#L172

it appears that chr() and unichr() are "supposed" to be builtin functions, while ord() is not mentioned.

In any case, all three of the above work fine in standard python (e.g. when run from the command line), but none of them work in the browser context.

What's worse, none of them reliably throw an error message.  On rare occasions I've seen a message of the form "ord is not defined" ... but usually nothing.  Execution of the compiled code stops abruptly, with no messages printed, and no exceptions thrown.

As a workaround, I wrote the following javascript:
```
    // Very rough, approximate, emergency workarounds for chr() and ord()
    function ord(ch){
      if (len(ch) > 1) return -1
      return ch.charCodeAt(0)
    }
    function chr(code){
      return String.fromCharCode(code)
    }
```

**Comment by BruceSherwood** [comment]
Clearly there's a problem with error reporting in the environment you're experimenting with. 

As for these functions, the issue is that RapydScript has an announced policy of not implementing Python functions that have adequate analogs in JavaScript. This is part of their stance of providing a "Python-like" environment for experienced web programmers. Happily, for a broad range of programs of the kind written by most GlowScript users, "Python-like" is indeed quite similar to true Python, but there are going to turn up cases like "ord" that have been left out of RapydScript. Such functions when identified are candidates for adding to GlowScript VPython. The upside of the RapydScript design philosophy is that by virtue of striving only to be "Python-like" rather emulating true Python as many other Python-to-JavaScript projects have attempted, RapydScript programs run about as fast as their JavaScript equivalents, whereas the others run very........slow.......ly.

**Comment by BruceSherwood** [comment]
I checked, and indeed when I try to use ord at glowscript.org I get the error message "ReferenceError: ord is not defined". If you want to make a stab at trying to connect to the glowscript.org error reporting, the main code is the function reportScriptError in untrusted/run.js.

**Comment by JohnDenker** [comment]
I understand the upside, but there is also a **major** downside to the pseudo-javascript policy:  It is  the enemy of portability.
  * Some people have an investment of many many thousands of lines of kosher python, and the aren't in the mood to rewrite it.
  * There are good reasons why somebody might want to run the same program in two different enviroments.  The web is great for disseminating something, but if you want to do file IO and/or make a movie, you really need to run in a more conventional, more privileged environment.

**Comment by BruceSherwood** [comment]
Handling legacy code has been my highest priority. 1) I installed RapydScript in the glowscript.org environment. 2) I extensively reworked both the compilation and execution libraries to implement the VPython API. 3) I provided a converter program written in true Python that does a good though necessarily not perfect job of converting an existing VPython program to run in a browser as well as running in classic VPython. 4) I made GlowScript VPython versions of 100 legacy programs, adjusting the GlowScript engine where necessary, as indicated by the conversion work. 5) I beefed up the error reporting so that for example box(color=(1,0,0)) gives a clear error message. This was a huge amount of work. Missing the ord() function is a nit, and had not previously been reported.

All of this said, there is no way that the browser VPython is ever going to be able to import numpy or PIL or thousands of other Python libraries, nor does it look likely that in the forseeable future it will be possible to get general operator overloading in a browser-based "Python". This means that it is simply not possible to do a perfect job, even for things much more important than the ord() function. For that reason there will always be a role for a VPython based on true Python, which gives access to the vast Python ecology of modules, because the browser version simply cannot and will not be able to do so. Several of us most concerned with the issue are committed to providing both true Python and browser versions of the VPython API.
